Add RKNN export to Benchmarks and Tests CI

lakshanthad opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved support for exporting YOLO models to the Rockchip RKNN format, with clearer documentation, better platform checks, and new automated tests. 🚀

### 📊 Key Changes
- Added a warning in the documentation about possible first-time export failures due to dependency installation for RKNN.
- Enabled RKNN export on both CPU and GPU in the exporter.
- Improved platform checks to ensure RKNN export only runs on supported systems (non-aarch64 Linux).
- Added a new automated test for RKNN export.
- Updated benchmarking logic to enforce correct platform and PyTorch version for RKNN export and inference.

### 🎯 Purpose & Impact
- 📝 **Clearer Guidance:** Users are now warned about potential first-time export issues, helping them troubleshoot more easily.
- 💻 **Better Compatibility:** Ensures RKNN export only runs where it's supported, reducing errors and confusion.
- 🧪 **Improved Reliability:** New tests help catch issues early and maintain export quality.
- ⚡ **Enhanced Usability:** Users working with Rockchip devices can expect a smoother export experience and clearer requirements.

lakshanthad committed:
Optimize RKNN export and inference structure

UltralyticsAssistant commented:
👋 Hello @lakshanthad, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a smooth review and integration process, please go through the following checklist:

- ✅ **Define a Purpose**: Ensure your PR description clearly explains the purpose of your changes, including any context or references to [relevant issues](https://github.com/ultralytics/ultralytics/issues). This helps maintainers quickly understand and evaluate your contribution.
- ✅ **Synchronize with Source**: Confirm that your branch is up-to-date with the `main` branch. You can do this by clicking the 'Update branch' button or running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Please verify that all applicable Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any CI jobs fail, review the logs and resolve the issues.
- ✅ **Update Documentation**: If your changes impact current functionality or add new features, ensure that the [documentation](https://docs.ultralytics.com/) is updated accordingly to reflect the adjustments.
- ✅ **Add Tests**: Include or revise tests as necessary to cover your changes. Make sure that all tests pass locally and in CI workflows.
- ✅ **Sign the CLA**: If this is your first contribution to Ultralytics, please sign our [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA."
- ✅ **Minimize Changes**: Keep your modifications scoped to the purpose mentioned in your PR description. Eliminate any unrelated or unnecessary changes for streamlined review and integration. Or, as Bruce Lee put it: _"Hack away the unessential."_ 🥋

For more details, don’t forget to check out our [Contributing Guide](https://docs.ultralytics.com/help/contributing/).

### Notes on your PR:
- The addition of RKNN export functionality 🎉 is a great enhancement for supporting Rockchip hardware use cases. Thank you!
- Please confirm that platform checks, such as `LINUX` and `ARM64`, properly cover all edge cases to ensure smooth usability.
- If not already done, ensure your changes are thoroughly tested on supported platforms to validate reliability. 🚀

An Ultralytics engineer will review this PR soon and provide additional feedback if necessary. In the meantime, feel free to reply here with any questions or additional details. Thanks again for your valuable contribution! 🙌

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
enable is_rockchip()

lakshanthad committed:
Add `rknn-toolkit2` to pyproject to fix dependency issue

lakshanthad committed:
unpin rknn-toolkit version

lakshanthad committed:
Update test_exports.py

lakshanthad committed:
remove `protobuf` pin

lakshanthad committed:
Update

lakshanthad committed:
More pin for `torch`

lakshanthad committed:
Update pyproject.toml

lakshanthad committed:
pin `protobuf`

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad commented on pyproject.toml:
```diff
@@ -68,7 +68,7 @@ dependencies = [
     "requests>=2.23.0",
     "scipy>=1.4.1",
-    "torch>=1.8.0",
+    "torch>=1.8.0,<=2.4.0", # rknn-toolkit2 dependency issue: https://github.com/ultralytics/ultralytics/pull/19482
```
@ambitious-octopus we cannot change versions in exporter because it will not pass CI in first run.

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Update pyproject.toml

lakshanthad committed:
Update exporter.py

lakshanthad committed:
Update test_exports.py

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Skip windows and mac

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Fix RKNN index

lakshanthad committed:
Disable test for arm64 Linux

lakshanthad committed:
Temporary disable rockchip check

lakshanthad committed:
Merge 3839dd6435e4a1d3f06c208c4f2e535c1a917b0a into 70dea780ace409117aa0088e0682d3769d2b8cf2

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

lakshanthad committed:
Update test_exports.py

lakshanthad committed:
Update benchmarks.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Update inference checks structure

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Update inference check logic

lakshanthad committed:
Update test_exports.py

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Make CPU and GPU True

lakshanthad committed:
Update exporter.py

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Remove pytorch constraints

lakshanthad committed:
Add torch constraint

lakshanthad committed:
Merge a491333b0e920bdbb121bd495135db929bb100d6 into 7b07f4a1a192e4ed6b2d092519c8e74bf9c7eb42

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

lakshanthad committed:
Add torch version check for RKNN export

lakshanthad committed:
Merge b7b6802751fffc52308c7b30c233dd8b431eb586 into 7b07f4a1a192e4ed6b2d092519c8e74bf9c7eb42

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

lakshanthad committed:
Add TODO comment for PyTorch version fixing

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Add RKNN export test to slow CI

lakshanthad committed:
Add extra line

lakshanthad commented:
@ultralytics/run-ci

lakshanthad committed:
Add export fail note in doc

lakshanthad commented:
@glenn-jocher This is good to go. Two points to note:

- Currently export test is implemented and passes because export can be performed on X86 Linux device. I will implement inference tests and benchmarks, once I receive the Rockchip devices and setup the CI runners.
- I have done an assertion for benchmarks CI because it fails in first run. This is due to `rknn-toolkit` downgrading PyTorch to 2.4.0. I have created an [issue on rknn-toolkit2 repo](https://github.com/airockchip/rknn-toolkit2/issues/362), emailed them and also added it as a note in this PR. However, it passes in second run because the downgraded PyTorch will be reflected there and I have added a note about this in the documentation.

The above points can be fixed and implemented in a future PR.

FYI: Slow CIs are passing [here](https://github.com/ultralytics/ultralytics/actions/runs/15232006812/job/42840892279)

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad commented:
@Laughing-q Good idea! I just tried `--no-deps` and it works locally. Not sure how we can use this when using `check_requirements` function.

lakshanthad committed:
Add `--no-deps` to skip dep install for `rknn-toolkit2`

lakshanthad committed:
Merge fecc05e9b4a1f3d14b9e3b665d3cffe803e460b9 into ac0b54b926a6976ba67ada43ca9eba36e1e7f6e5

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

lakshanthad commented:
@Laughing-q Testing with `--no-deps` now

lakshanthad committed:
Change export order

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Fix benchmarks format

lakshanthad committed:
Update benchmarks.py

lakshanthad committed:
Update exporter.py

lakshanthad committed:
Update exporter.py

lakshanthad committed:
Update exporter.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Revert "Update exporter.py"

This reverts commit 62c0ff633266691d52d8033bd8086f3691b31417.

lakshanthad committed:
Merge branch 'optimize-rknn-tests-benchmarks' of https://github.com/ultralytics/ultralytics into optimize-rknn-tests-benchmarks

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Update exporter.py

lakshanthad committed:
Merge branch 'optimize-rknn-tests-benchmarks' of https://github.com/ultralytics/ultralytics into optimize-rknn-tests-benchmarks

lakshanthad committed:
Fix RKNN export order

lakshanthad committed:
Update exporter.py

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Fix order for RKNN

lakshanthad committed:
Try placing rknn above ncnn

lakshanthad committed:
Place rknn after onnx

lakshanthad committed:
Place rknn back in original position

lakshanthad committed:
Enable assertion for exporter

lakshanthad committed:
Merge 8309d20fc7a4c704cb46d08e5646e62de2cfc2d6 into dae5ab441c99c08ccfa6077c7f76d0c7e925cd19

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

lakshanthad committed:
Remove `--no-deps` argument

lakshanthad committed:
Revert "Remove `--no-deps` argument"

This reverts commit b77c7ca79b5315f2cf3a8685364aa563da13baba.

lakshanthad committed:
Add comment in `rknn-toolkit2` installation

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad commented:
@Laughing-q I have tried experimenting with placing RKNN export much earlier in the order when benchmarking different export formats and also added `--no-deps` flag. However this failure is till there:

```
AssertionError: Benchmark failure for RKNN: (torch 2.7.0+cpu (/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages), Requirement.parse('torch<=2.4.0,>=1.10.1'), {'rknn-toolkit2'})
```

It seems that for the time being, the only viable solution is to do a `PyTorch<=2.4.0` assertion until Rockchip team fixes PyTorch version compatibility.

The latest commit reflects this change and all the CIs are passing.

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

Laughing-q committed:
revert rknn check

Laughing-q committed:
block other tests

Laughing-q committed:
add format=rknn

lakshanthad committed:
Merge 92750da2afeae21c56e78de5ccce00b4a42e9576 into d564179793b6e6579a08351e2a7d948fe19ad8ca

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

Laughing-q committed:
temporarily remove "path with spaces"

Laughing-q committed:
Update ci.yml

Laughing-q committed:
Update ci.yml

Laughing-q committed:
block other formats

Laughing-q committed:
revert other formats

Laughing-q committed:
Update exporter.py

Laughing-q committed:
add CoreML back

Laughing-q committed:
add tensorflow savedmodel

Laughing-q committed:
add formats

Laughing-q committed:
add tensorflow.js back

Laughing-q committed:
manually add tensorflowjs import

Laughing-q commented:
@lakshanthad ok as discussed, let me have a briefly summary here about what I found in this PR and let's hold this PR for now.
- Benchmark failing on rknn: because the `import tensorflowjs as tfjs` in previous `tfjs` benchmark tests somehow affecting `rknn-toolit2` installation(`check_requirements("rknn-toolkit2", cmds="--no-deps")`) hence the torch version conflict(2.7.0 in tests env and rknn package needs at least `<=2.4.0`) throws an error, even .
  ```bash
  AssertionError: Benchmark failure for RKNN: (torch 2.7.0+cpu (/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages), Requirement.parse('torch<=2.4.0,>=1.10.1'), {'rknn-toolkit2'})
  ```
  - noted that doing rknn benchmark independently would work, it's confirmed that somehow `tensorflowjs` importing affect the installation of rknn package, probably it imports some dependencies of rknn as well.
    ```bash
    yolo benchmark model=yolo11n.pt format=rknn
    ```
- Even found a way to bypass the issue above in tests, I found currently rknn package does not work with `path with spaces`, it'd fail if we export rknn format on models in a directory with spaces in its name.
  ```bash
  yolo export model="path with spaces/yolo11n.pt" format=rknn
  ```
  ```bash
  FileNotFoundError: [Errno 2] No such file or directory: 'yolo11n_rknn_model/path with spaces/yolo11n-rk3588.rknn'
  ```

@glenn-jocher For the first issue we've seen multiple times in our tests that other packages importing are affecting next packages installation, at least for ci tests we need to figure a way to install all export relevant dependencies at once.

Laughing-q committed:
roll back ci.yml

Laughing-q committed:
roll back tests spaces

Laughing-q committed:
Update exporter.py

Laughing-q committed:
Update ci.yml

Laughing-q committed:
Update exporter.py

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad commented:
Converted PR to draft and removed TODO @Laughing-q until this becomes ready to merge. We will have to wait until the above-mentioned issues are resolved by Rockchip team.

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

lakshanthad committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

Signed-off-by: Lakshantha Dissanayake <lakshantha@ultralytics.com>

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

UltralyticsAssistant committed:
Merge branch 'main' into optimize-rknn-tests-benchmarks

onuralpszr commented:
@lakshanthad since pypi is old and not update and conflicts are getting bigger, closing because of no update from "https://pypi.org/project/rknn-toolkit2/" so that change we can re-open so I am closing in favor "no update from rknn packages"
